chips/sam4l/src/aes.rs

```diff
@@ -0,0 +1,258 @@
+//! Implementation of the AESA peripheral on the SAM4L.
+
+use core::cell::Cell;
+use core::mem;
+
+use kernel::common::VolatileCell;
+use kernel::common::take_cell::TakeCell;
+use kernel::hil;
+use nvic;
+use pm;
+use scif;
+
+/// The registers used to interface with the hardware
+#[repr(C, packed)]
+struct AesRegisters {
+    ctrl: VolatileCell<u32>, //       0x00
+    mode: VolatileCell<u32>, //       0x04
+    databufptr: VolatileCell<u32>, // 0x08
+    sr: VolatileCell<u32>, //         0x0C
+    ier: VolatileCell<u32>, //        0x10
+    idr: VolatileCell<u32>, //        0x14
+    imr: VolatileCell<u32>, //        0x18
+    _reserved0: VolatileCell<u32>, // 0x1C
+    key0: VolatileCell<u32>, //       0x20
+    key1: VolatileCell<u32>, //       0x24
+    key2: VolatileCell<u32>, //       0x28
+    key3: VolatileCell<u32>, //       0x2c
+    key4: VolatileCell<u32>, //       0x30
+    key5: VolatileCell<u32>, //       0x34
+    key6: VolatileCell<u32>, //       0x38
+    key7: VolatileCell<u32>, //       0x3c
+    initvect0: VolatileCell<u32>, //  0x40
+    initvect1: VolatileCell<u32>, //  0x44
+    initvect2: VolatileCell<u32>, //  0x48
+    initvect3: VolatileCell<u32>, //  0x4c
+    idata: VolatileCell<u32>, //      0x50
+    _reserved1: [u32; 3], //          0x54 - 0x5c
+    odata: VolatileCell<u32>, //      0x60
+    _reserved2: [u32; 3], //          0x64 - 0x6c
+    drngseed: VolatileCell<u32>, //   0x70
+}
+
+// Section 7.1 of datasheet
+const AES_BASE: u32 = 0x400B0000;
+
+pub struct Aes {
+    registers: *mut AesRegisters,
```

Comment:
I think is more idiomatic Rust to de-reference instead of transmuting consider to use the following instead (optional imo):
registers: *const AesRegisters
regs: AES_BASE as *const AesRegisters
let registers = unsafe { &*self.registers };